src/votekit/ballot_graph.py

```diff
@@ -0,0 +1,226 @@
+import matplotlib.pyplot as plt
+import networkx as nx
+from collections import Counter
+from profile import PreferenceProfile
+from distinctipy import get_colors
+from ballot import Ballot
+from typing import Optional, Callable
+from pydantic import BaseModel, validator
+
+
+def build_graph(n):
+
+    Gc = nx.Graph()
+
+        # base cases
+    if n==1:
+        Gc.add_nodes_from([(1)])
+    elif n==2:
+        Gc.add_nodes_from([(1,2),(2,1)])
+        Gc.add_edges_from([((1,2), (2,1))])
+
+    else:
+            
+        if n-1 not in Graphs.keys(): 
+            # make the adjacency graph of size (n - 1)
+            build_graph(n-1)
+        G_prev = Graphs[n-1]
+        for i in range(1,n+1):
+                # add the node for the bullet vote i
+            Gc.add_node(tuple([i]))
+
+            # make the subgraph for the ballots where i is ranked first
+            G_corner = relabel(G_prev,i,n)
+      
+                # add the components from that graph to the larger graph
+            Gc.add_nodes_from(G_corner.nodes)
+            Gc.add_edges_from(G_corner.edges)
+
+                # connect the bullet vote node to the appropriate verticies
+            if n == 3:
+                Gc.add_edges_from([(k,tuple([i])) for k in G_corner.nodes])
+            else:
+                Gc.add_edges_from([(k,tuple([i])) for k in G_corner.nodes if len(k) == 2])
+            
+        nodes = Gc.nodes
+
+            # add the additional edges corresponding to swapping the order of the
+            # first two candidates
+        new_edges = []
+        for k in nodes:
+            if len(k)==2:
+                new_edges.append(((k[0],k[1]),(k[1],k[0])))
+            elif len(k)>2:
+                l = list(k)
+                a = l[0]
+                b=l[1]
+                new_edges.append((tuple([a]+[b]+l[2:]), tuple([b]+[a]+l[2:])))
+
+        Gc.add_edges_from(new_edges)
+
+    Graphs[n] = Gc
+    return Gc
+
+
+def relabel(gr, new_label, num_cands):
+    node_map = {}
+    graph_nodes = list(gr.nodes)
+
+
+    for k in graph_nodes:
+            # add the value of new_label to every entry in every ballot
+        tmp = [new_label+y for y in k]
+
+            # reduce everything mod new_label
+        for i in range(len(tmp)):
+            if tmp[i]> num_cands:
+                tmp[i] = tmp[i]- num_cands
+        node_map[k] = tuple([new_label]+tmp)
+
+    return nx.relabel_nodes(gr, node_map)
+
+
+Graphs = {}
```

Comment:
this global variable seems to be used for caching, it should either be labeled with a comment justifying its usage or replaced with something intended for caching (see functools.lru_cache)

Reply:
Deleted the variable Graphs

```suggestion
@cache
def build_graph(n):
    print("Called on" + str(n))
    Gc = nx.Graph()
```